R/run-nmtran.R

```diff
@@ -0,0 +1,246 @@
+        stats::setNames(names(input_data))
+    }, error = function(cond){
+      rlang::inform(
+        c("FDATA could not be read in:", cond$parent$message)
+      )
+      return(NULL)
+    })
+
+    if(!is.null(fdata)){
+      attr(fdata, "n_records_dropped") <- nrow(input_data) - nrow(fdata)
+    }
+
+    # assign class and return
+    class(fdata) <- c(NMTRAN_FDATA_CLASS, class(fdata))
+    return(fdata)
+  }else{
+    return(invisible(NULL))
+  }
+}
+
+
+#' Search for and validate existence of an `NM-TRAN` executable
+#'
+#' If `nmtran_exe = NULL`, this will look for a `bbi.yaml` file in the same
+#' directory as the model.
+#'
+#' @inheritParams run_nmtran
+#'
+#' @keywords internal
+locate_nmtran <- function(.mod = NULL, .config_path = NULL, nmtran_exe = NULL){
+
+  if(is.null(nmtran_exe)){
+    if(!is.null(.mod)){
+      check_model_object(.mod, "bbi_nonmem_model")
+      model_dir <- get_model_working_directory(.mod)
+    }
+    config_path <- .config_path %||% file.path(model_dir, "bbi.yaml")
+
+    if(!file_exists(config_path)){
+      rlang::abort(
+        c(
+          "x" = "No bbi configuration was found in the execution directory.",
+          "i" = "Please run `bbi_init()` with the appropriate directory to continue."
+        )
+      )
+    }
+
+    if(!is.null(.config_path)){
+      config_path <- normalizePath(.config_path)
+    }
+
+    bbi_config <- yaml::read_yaml(config_path)
+    nm_config <- bbi_config$nonmem
+
+    # look for default nonmem installation
+    default_nm <- purrr::keep(nm_config, function(nm_ver){
+      !is.null(nm_ver$default)
+    })
+
+    # Set nonmem path
+    if(length(default_nm) > 0){
+      default_nm <- default_nm[[1]]
+    }else{
+      # If no default, use the last one (likely higher version)
+      default_nm <- nm_config[[length(nm_config)]]
+    }
+
+    # Set NMTRAN executable path
+    nm_path <- default_nm$home
+    # TODO: should we recursively look for this executable, or assume Metworx?
+    # i.e. can we assume NMTRAN is in a `tr` folder, and is called `NMTRAN.exe`?
+    nmtran_exe <- file.path(nm_path, "tr", "NMTRAN.exe")
+
+    # If executable found via bbi.yaml, append NONMEM version as attribute
+    attr(nmtran_exe, "nonmem_version") <- basename(default_nm$home)
+  }
+
+  if(!file_exists(nmtran_exe)){
+    stop(glue("Could not find an NMTRAN executable at `{nmtran_exe}`"))
+  }
+
+  return(nmtran_exe)
+}
+
+
+#' Execute `NM-TRAN` in a given directory
+#'
+#' @param nmtran_exe Path to `NM-TRAN` executable.
+#' @param mod_path Path of a model to evaluate. Should be relative to `dir`.
+#' @param dir Directory in which to execute the command.
+#'
+#' @keywords internal
+execute_nmtran <- function(nmtran_exe, mod_path, dir = NULL) {
+  if(is.null(dir)) dir <- "."
+  checkmate::assert_directory_exists(dir)
+
+  nmtran.p <- processx::process$new(
+    command = nmtran_exe, args = mod_path, wd = dir,
+    stdout = "|", stderr="|", stdin = file.path(dir, mod_path)
+  )
```

Comment:
* Did you look into the questions a brought up at <https://github.com/metrumresearchgroup/bbr/issues/650#issuecomment-1930725642>?
 * Why is `mod_path` passed as `args`?

Reply:
Some notes related to this:

 * The call to `nmtran_presort` starts with nmfe74.  nmfe73 doesn't call `nmtran_presort`.

 * Only nmfe75 does the upfront NMTRAN call with the do2test argument

   * bbi does not expose the do2test argument

   * If the exit status of that code is 9, it does a follow-up call without passing the do2test arg

   * For the purpose of `run_nmtran`, it's probably safe to just ignore this upfront call given bbi doesn't expose the argument

 * The arguments of the main NMTRAN call are prdefault, tprdefault, and maxlim.  bbi exposes prdefault and tprdefault as of v3.2.0.  It has exposed maxlim for much longer, but in v3.2.0 it started passing `maxlim=2` by default.

   * These options can be set by `nmfe_options` in bbi.yaml and via `.bbi_args`

 * Starting with nmfe75, the NMTRAN pipe begins with the contents of `tempzzzz1` on stdin instead of `$1` (the original control stream).  This looks to be relevant only when `-dde` is passed.  bbi does not expose that flag.  So, from our standpoint, it's fine to take `tempzzzz1` as equivalent to the original control stream.

---

Initial thoughts:

 * While I don't know what `nmtran_presort` is doing, it seems to me that the safe option would be to feed the input through it.  We can handle the nmfe73 case by skipping this step if `util/nmtran_presort` doesn't exist.

 * For the NMTRAN arguments: at the very least, I think we should 1) document what values for prdefault, tprdefault, and maxlim are used and 2) pass arguments matching what would be triggered by a vanilla bbi (`"0" "0" "2"`, I think, but will verify).

   It'd be better to honor `bbi.yaml` and probably even expose arguments to `run_nmtran`, but I don't think it's critical so long as `run_nmtran` is exposed as a standalone helper rather than wired up to `submit_model` as a default preflight check.

Reply:
>Why is mod_path passed as args

This is the only way I could get the `NM-TRAN` command to work with `processx`. I tried a number of alternatives (not specifying `stdin` or `wd`, changing the command, etc) and they would always error out with a message similar to this:
```r
# Tried variants of the expression below
nmtran.p <- processx::process$new(
  command = "/opt/NONMEM/nm74gf/tr/NMTRAN.exe < 100.ctl", 
  stdout = "|", stderr="|", stdin = file.path(dir, mod_path)
)

Error in c("process_initialize(self, private, command, args, stdin, stdout, ",  : 
  ! Native call to `processx_exec` failed
Caused by error in `chain_call(c_processx_exec, command, c(command, args), pty, pty_options, …` at initialize.R:138:3:
! cannot start processx process '/opt/NONMEM/nm74gf/tr/NMTRAN.exe < 100.ctl' (system error 2, No such file or directory) @unix/processx.c:613 (processx_exec)
```

Reply:
Sorry I deleted the comment you referenced because I figured it out (must have set things up incorrectly one time)

Reply:
@kyleam you mentioned earlier that `nmfe_options` can be defined in the `bbi.yaml`, and they can also be passed via `bbi_args()` to other functions. Should I add a `bbi_args` argument and attempt to pull those options (if specified) from there, and otherwise parse from `bbi.yaml`/use the defaults?

Reply:
> Should I add a `bbi_args` argument and attempt to pull those options [...]

From a functionality standpoint, I think that (i.e. trying to mimic what the behavior would be on a submit_model call, including the bbi-side logic) would be the most correct approach.  However, I'd also be okay with the minimal approach I mentioned above:

>> document what values for prdefault, tprdefault, and maxlim are used and 2) pass arguments matching what would be triggered by a vanilla bbi ("0" "0" "2", I think, but will verify)

I'd say that's good enough, at least for the initial implementation.  This is standalone functionality that doesn't affect other things, so a discrepancy between  the `run_nmtran` NMTRAN output/exit and the exit/output of the NMTRAN call triggered by bbi/nmfe is unfortunate but doesn't bring things crashing down.

So, I don't have a strong opinion and am okay with whatever you prefer.  I think the three options are 1) implement minimum approach and leave more involved one be for now, 2) implement minimum approach and plan to follow up with a separate PR to switch to more involved approach, or 3) implement more involved approach in this PR.